### PR TITLE
Subscription results keep in upstream order

### PR DESCRIPTION
**Pull request by bbakerman**
See https://github.com/spring-projects/spring-graphql/issues/949

This PR has a reproduction of the problem reported and also a series fo fixes

However after discussions with the Spring team, its not apparent that the old "stream what completes first" approach is wrong.

So I have added two modes - the older `CompletionStageMappingPublisher` which will emit objects as soon as they complete and a newer `CompletionStageMappingOrderedPublisher` that will emit elements in the original upstream Publisher order.

Both have bufferring (for elements that have not completed when  they are mapped into graphql objects) but once keeps them in original presentation order while the old one does what it has always done.

I have split the inner Subscriber classes out into their own classes and hence can unit test them better

I have changed the tests considerable.  I have tested more edge cases.

I have also made them finish up when `onError` is called and they cancel any futures in flight.

### CODE REVIEW

**Comment by bbakerman** [comment]
Updated it to have a `Mono` property thats async

```
onSubscribe
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=0, name=name0, lastEpisode=episode-boundedElastic-1for0, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=1, name=name1, lastEpisode=episode-boundedElastic-1for1, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=2, name=name2, lastEpisode=episode-boundedElastic-1for2, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=3, name=name3, lastEpisode=episode-boundedElastic-1for3, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=4, name=name4, lastEpisode=episode-boundedElastic-1for4, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=5, name=name5, lastEpisode=episode-boundedElastic-1for5, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=6, name=name6, lastEpisode=episode-boundedElastic-2for6, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=7, name=name7, lastEpisode=episode-boundedElastic-2for7, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=8, name=name8, lastEpisode=episode-boundedElastic-2for8, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=9, name=name9, lastEpisode=episode-boundedElastic-1for9, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=10, name=name10, lastEpisode=episode-boundedElastic-1for10, isFavorite=true}}, dataPresent=true, extensions=null}
complete

```

**Comment by madwind** [comment]
when request one by one

```java
    private CompletableFuture<Flux<Object>> mkFluxDF(DataFetchingEnvironment env) {
        // async deliver of the publisher with random snoozing between values
        return CompletableFuture.supplyAsync(() -> Flux.generate(() -> 0, (counter, sink) -> {
            sink.next(mkValue(counter));
            snooze(rand(10, 100));
            if (counter == 10) {
                sink.complete();
            }
            return counter + 1;
        }).doOnNext(System.out::println));
    }
```
```
onSubscribe
{id=0, counter=0}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=0, name=name0, lastEpisode=episode-boundedElastic-1for0, isFavorite=true}}, dataPresent=true, extensions=null}
{id=1, counter=1}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=1, name=name1, lastEpisode=episode-boundedElastic-2for1, isFavorite=false}}, dataPresent=true, extensions=null}
{id=2, counter=2}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=2, name=name2, lastEpisode=episode-boundedElastic-2for2, isFavorite=true}}, dataPresent=true, extensions=null}
{id=3, counter=3}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=3, name=name3, lastEpisode=episode-boundedElastic-2for3, isFavorite=false}}, dataPresent=true, extensions=null}
{id=4, counter=4}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=4, name=name4, lastEpisode=episode-boundedElastic-2for4, isFavorite=true}}, dataPresent=true, extensions=null}
{id=5, counter=5}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=5, name=name5, lastEpisode=episode-boundedElastic-1for5, isFavorite=false}}, dataPresent=true, extensions=null}
{id=6, counter=6}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=6, name=name6, lastEpisode=episode-boundedElastic-1for6, isFavorite=true}}, dataPresent=true, extensions=null}
{id=7, counter=7}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=7, name=name7, lastEpisode=episode-boundedElastic-1for7, isFavorite=false}}, dataPresent=true, extensions=null}
{id=8, counter=8}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=8, name=name8, lastEpisode=episode-boundedElastic-2for8, isFavorite=true}}, dataPresent=true, extensions=null}
{id=9, counter=9}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=9, name=name9, lastEpisode=episode-boundedElastic-2for9, isFavorite=false}}, dataPresent=true, extensions=null}
{id=10, counter=10}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=10, name=name10, lastEpisode=episode-boundedElastic-2for10, isFavorite=true}}, dataPresent=true, extensions=null}
complete
```

when request 10
```java
        @Override
        public void onSubscribe(Subscription subscription) {
            this.subscription = subscription;
            System.out.println("onSubscribe");
            subscription.request(10);
        }

        @Override
        public void onNext(Object o) {
            System.out.println("\tonNext : " + o);
//            subscription.request(1);
        }
```
{id=0, counter=0}
{id=1, counter=1}
{id=2, counter=2}
{id=3, counter=3}
{id=4, counter=4}
{id=5, counter=5}
{id=6, counter=6}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=0, name=name0, lastEpisode=episode-boundedElastic-1for0, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=2, name=name2, lastEpisode=episode-boundedElastic-3for2, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=3, name=name3, lastEpisode=episode-boundedElastic-4for3, isFavorite=false}}, dataPresent=true, extensions=null}
{id=7, counter=7}
{id=8, counter=8}
{id=9, counter=9}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=1, name=name1, lastEpisode=episode-boundedElastic-2for1, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=4, name=name4, lastEpisode=episode-boundedElastic-5for4, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=5, name=name5, lastEpisode=episode-boundedElastic-1for5, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=6, name=name6, lastEpisode=episode-boundedElastic-2for6, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=8, name=name8, lastEpisode=episode-boundedElastic-4for8, isFavorite=true}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=7, name=name7, lastEpisode=episode-boundedElastic-3for7, isFavorite=false}}, dataPresent=true, extensions=null}
	onNext : ExecutionResultImpl{errors=[], data={searchVideo={id=9, name=name9, lastEpisode=episode-boundedElastic-5for9, isFavorite=false}}, dataPresent=true, extensions=null}
```

```

File: src/main/java/graphql/execution/SubscriptionExecutionStrategy.java [hunk 1]
```diff
@@ -37,6 +38,14 @@
...
 public class SubscriptionExecutionStrategy extends ExecutionStrategy {
 
+    /**
+     * If a boolean value is placed into the {@link GraphQLContext} with this key then the order
+     * of the subscription events can be controlled.   By default, subscription events are published
+     * as the graphql subselection calls complete, and not in the order they originally arrived from the
+     * source publisher.  But this can be changed to {@link Boolean#TRUE} to keep them in order.
+     */
+    public static final String KEEP_SUBSCRIPTION_EVENTS_ORDERED = "KEEP_SUBSCRIPTION_EVENTS_ORDERED";
+
```
Comment (bbakerman) [inline src/main/java/graphql/execution/SubscriptionExecutionStrategy.java:48]: By default this is off (old behavior) but it can be opted into

File: src/main/java/graphql/execution/reactive/CompletionStageMappingOrderedPublisher.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * A reactive Publisher that bridges over another Publisher of `D` and maps the results
+ * to type `U` via a CompletionStage, handling errors in that stage but keeps the results
+ * in order of downstream publishing. This means it must queue unfinished
+ * completion stages in memory in arrival order.
+ *
+ * @param <D> the downstream type
+ * @param <U> the upstream type to be mapped to
+ */
+@Internal
+public class CompletionStageMappingOrderedPublisher<D, U> extends CompletionStageMappingPublisher<D, U> {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageMappingOrderedPublisher.java:21]: it extends `CompletionStageMappingPublisher` because they have a LOT in common - just the ordering on inflight objects is different

Comment (bbakerman) [reply src/main/java/graphql/execution/reactive/CompletionStageMappingOrderedPublisher.java:21]: Note also this has been split out of the publisher class

File: src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java [hunk 3]
```diff
@@ -1,31 +1,26 @@
...
+ * @param <D> the downstream type
+ * @param <U> the upstream type to be mapped to
  */
-@SuppressWarnings("ReactiveStreamsPublisherImplementation")
 @Internal
 public class CompletionStageMappingPublisher<D, U> implements Publisher<D> {
-    private final Publisher<U> upstreamPublisher;
-    private final Function<U, CompletionStage<D>> mapper;
+    protected final Publisher<U> upstreamPublisher;
+    protected final Function<U, CompletionStage<D>> mapper;
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java:23]: not designed to be inherited from its ordered peer

File: src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * This subscriber can be used to map between a {@link org.reactivestreams.Publisher} of U
+ * elements and map them into {@link CompletionStage} of D promises, and it keeps them in the order
+ * the Publisher provided them.
+ *
+ * @param <U> published upstream elements
+ * @param <D> mapped downstream values
+ */
+@Internal
+public class CompletionStageOrderedSubscriber<U, D> extends CompletionStageSubscriber<U, D> implements Subscriber<U> {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java:20]: It inherents from CompletionStageSubscriber because again they are very much the same

File: src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            }
+        });
+    }
+
+    private Throwable cfExceptionUnwrap(Throwable throwable) {
+        if (throwable instanceof CompletionException & throwable.getCause() != null) {
+            return throwable.getCause();
+        }
+        return throwable;
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java:82]: the onError will be the underling exception not the wrapper CompletionException that comes from CF usage

File: src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+/**
+ * This subscriber can be used to map between a {@link org.reactivestreams.Publisher} of U
+ * elements and map them into {@link CompletionStage} of D promises.
+ *
+ * @param <U> published upstream elements
+ * @param <D> mapped downstream values
+ */
+@Internal
+public class CompletionStageSubscriber<U, D> implements Subscriber<U> {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java:25]: This was split out from the peer Publisher code as its own file.  Its complicated enough to deserve that.

File: src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    protected final Function<U, CompletionStage<D>> mapper;
+    protected final Subscriber<? super D> downstreamSubscriber;
+    protected Subscription delegatingSubscription;
+    protected final Queue<CompletionStage<?>> inFlightDataQ;
+    protected final LockKit.ReentrantLock lock = new LockKit.ReentrantLock();
+    protected final AtomicReference<Runnable> onCompleteRun;
+    protected final AtomicBoolean isTerminal;
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java:32]: Designed to be inheret from by `CompletionStageOrderedSubscriber`

File: src/test/groovy/graphql/execution/SubscriptionExecutionStrategyTest.groovy [hunk 8]
```diff
@@ -628,4 +628,97 @@ class SubscriptionExecutionStrategyTest extends Specification {
...
+        GraphQL graphQL = buildSubscriptionQL(newMessageDF, senderDF, textDF)
+
+        def executionInput = ExecutionInput.newExecutionInput().query("""
+            subscription NewMessages {
+              newMessage(roomId: 123) {
+                sender
+                text
+              }
+            }
+        """).graphQLContext([(SubscriptionExecutionStrategy.KEEP_SUBSCRIPTION_EVENTS_ORDERED): true]).build()
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/SubscriptionExecutionStrategyTest.groovy:689]: our hint to make them ordered.  Here is an example

File: src/test/groovy/graphql/execution/reactive/CompletionStageMappingOrderedPublisherTest.groovy [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+import org.awaitility.Awaitility
+import org.reactivestreams.Publisher
+import spock.lang.Specification
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionException
+import java.util.concurrent.CompletionStage
+import java.util.function.Function
+
+class CompletionStageMappingOrderedPublisherTest extends Specification {
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/reactive/CompletionStageMappingOrderedPublisherTest.groovy:15]: The tests for the two classes kinda repeat themselves - but I figuered that even though they inherent from each other - the tests should stress them the same and make no assumptions on how they are implemented

With even more work maybe the tests become common ones use `where:` but I didnt go that far - copy pasta rather

File: src/test/groovy/graphql/execution/reactive/CompletionStageOrderedSubscriberTest.groovy [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+import graphql.execution.pubsub.CapturingSubscription
+import spock.lang.Specification
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+import java.util.function.Function
+
+import static graphql.execution.reactive.CompletionStageSubscriberTest.mapperThatDoesNotComplete
+
+class CompletionStageOrderedSubscriberTest extends Specification {
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/reactive/CompletionStageOrderedSubscriberTest.groovy:13]: by breaking out the Subscriber classes into their own class - we can unit test them better

File: src/test/groovy/graphql/execution/reactive/tck/CompletionStageMappingOrderedPublisherRandomCompleteTckVerificationTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
+package graphql.execution.reactive.tck;
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/reactive/tck/CompletionStageMappingOrderedPublisherRandomCompleteTckVerificationTest.java:1]: I have made a series of new TestNG compliance tests that use the TCK to ensure they are valid publishers

File: src/test/java/reproductions/SubscriptionReproduction.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+import static graphql.schema.idl.RuntimeWiring.newRuntimeWiring;
+import static graphql.schema.idl.TypeRuntimeWiring.newTypeWiring;
+
+/**
+ * Related to <a href="https://github.com/spring-projects/spring-graphql/issues/949">...</a>
+ * <p>
+ * This reproduction is to see what's happening with Subscriptions and whether they keep their
+ * order when values are async.
+ */
+public class SubscriptionReproduction {
```
Comment (bbakerman) [inline src/test/java/reproductions/SubscriptionReproduction.java:35]: This is a `main()` reproduction class - not a test - I wanted to find out  what was happening and I thought this was a good way to find out.  Other code now tests that the fix is in place

File: src/test/groovy/graphql/execution/reactive/CompletionStageMappingOrderedPublisherTest.groovy (hunk 9 above)
Comment (bbakerman) [reply src/test/groovy/graphql/execution/reactive/CompletionStageMappingOrderedPublisherTest.groovy]: And - no I ended up refactoring so that the same tests are run for both implementations - one in order and one not in order

File: build.gradle [hunk 13]
```diff
@@ -117,6 +117,7 @@ dependencies {
 
     testImplementation 'org.reactivestreams:reactive-streams-tck:' + reactiveStreamsVersion
     testImplementation "io.reactivex.rxjava2:rxjava:2.2.21"
+    testImplementation "io.projectreactor:reactor-core:3.6.5"
```
Comment (bbakerman) [inline build.gradle:120]: Added so we can do Reactor specific tests if we choose

File: src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java [hunk 14]
```diff
@@ -40,9 +35,16 @@ public CompletionStageMappingPublisher(Publisher<U> upstreamPublisher, Function<
 
     @Override
     public void subscribe(Subscriber<? super D> downstreamSubscriber) {
-        upstreamPublisher.subscribe(new CompletionStageSubscriber(downstreamSubscriber));
+        assertNotNullWithNPE(downstreamSubscriber, () -> "Subscriber passed to subscribe must not be null");
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java:38]: spec says it MUST be non null

File: src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java [hunk 15]
```diff
@@ -40,9 +35,16 @@ public CompletionStageMappingPublisher(Publisher<U> upstreamPublisher, Function<
...
+        upstreamPublisher.subscribe(createSubscriber(downstreamSubscriber));
+    }
+
+    @NotNull
+    protected Subscriber<? super U> createSubscriber(Subscriber<? super D> downstreamSubscriber) {
+        return new CompletionStageSubscriber<>(mapper, downstreamSubscriber);
     }
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java:45]: The ordering comes from the Subscriber impl

File: src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java [hunk 16]
```diff
@@ -52,126 +54,4 @@ public Publisher<U> getUpstreamPublisher() {
...
-            //
-            delegatingSubscription.cancel();
-        }
-
-        @Override
-        public void onError(Throwable t) {
-            onCompleteOrError(() -> {
-                onCompleteOrErrorRunCalled.set(true);
-                downstreamSubscriber.onError(t);
-            });
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageMappingPublisher.java:130]: See in the old impl we waited until ALL inflights finished before we called onError - well now we dont.

We fail fast after discussing it with the Spring folks

File: src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java (hunk 4 above)
Comment (bbakerman) [reply src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java:20]: except for ordering of source events

File: src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+            // from the top of the in flight queue, take all the CFs that have
+            // completed... but stop if they are not done
+            while (!inFlightDataQ.isEmpty()) {
+                CompletionStage<?> cs = inFlightDataQ.peek();
+                if (cs != null) {
+                    //
+                    CompletableFuture<?> cf = cs.toCompletableFuture();
+                    if (cf.isDone()) {
+                        // take it off the queue
+                        inFlightDataQ.poll();
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageOrderedSubscriber.java:55]: in a lock so we can rely on it not changing here on the second mutative read

File: src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    @Override
+    public void onError(Throwable t) {
+        // we immediately terminate - we don't wait for any promises to complete
+        if (isTerminal.compareAndSet(false, true)) {
+            downstreamSubscriber.onError(t);
+            cancelInFlightFutures();
+        }
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java:140]: now fails fast

File: src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+    @Override
+    public void onComplete() {
+        onComplete(() -> {
+            if (isTerminal.compareAndSet(false, true)) {
+                downstreamSubscriber.onComplete();
+            }
+        });
```
Comment (bbakerman) [inline src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java:148]: waits for all inflight CFs to complete

File: src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+    protected Subscription delegatingSubscription;
+    protected final Queue<CompletionStage<?>> inFlightDataQ;
+    protected final LockKit.ReentrantLock lock = new LockKit.ReentrantLock();
+    protected final AtomicReference<Runnable> onCompleteRun;
+    protected final AtomicBoolean isTerminal;
+
+    public CompletionStageSubscriber(Function<U, CompletionStage<D>> mapper, Subscriber<? super D> downstreamSubscriber) {
+        this.mapper = mapper;
+        this.downstreamSubscriber = downstreamSubscriber;
+        inFlightDataQ = new ArrayDeque<>();
```
Comment (dondonz) [inline src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java:37]: Nitpick: to a non-native English speaker it's going to be confusing what the "Q" is (it being a homophone for "queue"). I can see it's not introduced in this PR, it's based off naming from previous code.

Could rename these methods/variables to "Queue" or mention in a comment that the "Q" means queue.
